In [1]:
import time
start_time = time.time()

In [2]:
# Importing the Libraries
import pandas as pd
import numpy as np
import tensorflow as tf
import keras
import random
import re

In [3]:
# Reading the files from the folder

fake_news_df = pd.read_csv('Fake.csv')
true_news_df = pd.read_csv('True.csv')

In [4]:
# Check the random instances of the data

display(fake_news_df.sample(2), true_news_df.sample(2))

title  \
8271    Superstar Marc Anthony Rocks Madison Square G...   
22802  Liberal Imperium: Quigley’s Anglo-American Est...   

                                                    text      subject  \
8271   If Donald Trump had been at Madison Square Gar...         News   
22802  Jay Dyer 21st Century WireThis is an addendum ...  Middle-east   

                   date  
8271   February 7, 2016  
22802   August 25, 2017

title  \
21053  Kenya to hold new presidential vote on Oct. 17...   
6335   U.S. House clears way for vote to start Obamac...   

                                                    text       subject  \
21053  NAIROBI (Reuters) - Kenya will hold a new pres...     worldnews   
6335   WASHINGTON (Reuters) - A majority of the U.S. ...  politicsNews   

                     date  
21053  September 4, 2017   
6335    January 13, 2017

## Data Cleaning Functions

In [5]:
## Data Cleaning  ###

# Remove the HTML text/phases from the data
def remove_html(text):
    new_text = re.sub(r'''(?i)\b((?:https?://|www\d{0,3}[.]|[a-z0-9.\-]+[.][a-z]{2,4}/)(?:[^\s()<>]+|\(([^\s()<>]+|(\([^\s()<>]+\)))*\))+(?:\(([^\s()<>]+|(\([^\s()<>]+\)))*\)|[^\s`!()\[\]{};:'".,<>?«»“”‘’]))''', "", text)
    return(new_text)

# Count the lenght of the string
def len_text(text):
    text_len = len(text.split())
    return(text_len)
    
# Remove White Spaces
def remove_white_space(text):
    text = re.sub("^\s+|\s+$", "", text, flags=re.UNICODE) # Remove spaces both in beginining and in the end of a string
    text = " ".join(re.split("\s+", text, flags=re.UNICODE)) # Remove spaces from duplicate spaces
    return(text)

## Data Analysis (Before Data Cleaning)

In [6]:
display(fake_news_df.sample(2), true_news_df.sample(2))

title  \
9564  HOUSE OF LIES: Pedophilia Allegations Levied A...   
4101   GOP Lawmaker: ‘A Lady Needs To Be Told When S...   

                                                   text   subject  \
9564  How should the gay community feel about the Tr...  politics   
4101  A Texas lawmaker just defended Donald Trump s ...      News   

                  date  
9564      Oct 30, 2017  
4101  October 21, 2016

title  \
18751  IAEA says North Korea's rapid weapons progress...   
15197  Philippines' Duterte offers to host 'world sum...   

                                                    text    subject  \
18751  SEOUL (Reuters) - The United Nations nuclear w...  worldnews   
15197  MANILA (Reuters) - Philippine President Rodrig...  worldnews   

                      date  
18751  September 29, 2017   
15197   November 10, 2017

##### FAKE News

In [7]:
display(set(fake_news_df['subject']), set(true_news_df['subject']))

{'Government News', 'Middle-east', 'News', 'US_News', 'left-news', 'politics'}

{'politicsNews', 'worldnews'}

In [8]:
# Get the length of each instance
fake_news_df['len_sent'] = fake_news_df['text'].apply(lambda x: len_text(x)) 

In [9]:
fake_news_df['subject'].value_counts()

News               9050
politics           6841
left-news          4459
Government News    1570
US_News             783
Middle-east         778
Name: subject, dtype: int64

In [10]:
# Analyse the description of each group in the subject feature
fake_news_df.groupby(['subject']).describe()

len_sent                                                       \
                   count        mean          std   min     25%    50%    75%   
subject                                                                         
Government News   1570.0  387.404459   389.428324   0.0  155.00  304.5  506.0   
Middle-east        778.0  772.548843  1040.851331  24.0  182.25  351.5  919.5   
News              9050.0  441.253812   152.060179  36.0  344.00  410.0  506.0   
US_News            783.0  780.527458  1049.778528  24.0  183.00  355.0  927.5   
left-news         4459.0  392.736264   363.389015   0.0  186.50  318.0  509.5   
politics          6841.0  346.752083   369.246542   0.0  120.00  276.0  462.0   

                         
                    max  
subject                  
Government News  4547.0  
Middle-east      8135.0  
News             3909.0  
US_News          8135.0  
left-news        7033.0  
politics         7033.0

#### Conclusion for Statistical analysis on Fake News Dataframe

- Maximum instances of possess by "News" with 9050 instances, followed by "politics" [6841], "left-news" [4459], "Govt News" [1570], "US_news" [783], and "Middle-east" [778]
- The news corpus is largely aligned towards the "left-news" and "politics" 
- The minimum number of text in Middle-east, News and US_News starts from 24, 36 and 24, respectively.
- There are various instances in Govt News, left-news and politics whether the news is empty

In [ ]:
random.seed(123)
fake_news_df[fake_news_df['subject']=='politics']['text'].sample(2, random_state=123).to_list()

## Data Analysis (After Data Cleaning)

In [ ]:
fake_news_df.head(1)

In [ ]:
fake_news_df['text'] = fake_news_df['text'].apply(lambda x: remove_html(x))
fake_news_df['text'] = fake_news_df['text'].apply(lambda x: remove_white_space(x))
fake_news_df['len_sent'] = fake_news_df['text'].apply(lambda x: len_text(x)) 

In [ ]:
fake_news_df.sample(2)

In [ ]:
## Analysis on Sentence Length
fake_news_df.describe()

In [ ]:
# Check the number of sentences below 10 word length
fake_news_df[fake_news_df['len_sent'] < 21].sort_values(by="len_sent",ascending=False)

In [ ]:
fake_news_df = fake_news_df[fake_news_df['len_sent'] > 20].reset_index(drop=True)
display(fake_news_df.shape, fake_news_df.sample(2))

In [ ]:
print("--- %s seconds ---" % (time.time() - start_time))